### DISCUSSION TITLE
Replicas for a Microservice doesn't allow string interpolation

### DISCUSSION THREAD

**Discussion by rmuller-vu** [mia-platform/community#543] [General]
### Description

The replicas property for any Microservice causes Mia to fail and generate the config to send to the orchestration generator.
There is a message that states string interpolation is allowed. Is this expected to allow interpolation?

Property Message
![image](https://github.com/user-attachments/assets/8eb3fc4e-6750-42c8-b879-9d5c6717d744)

Error Message on Deploy Screen
![image](https://github.com/user-attachments/assets/2c240e06-de4e-4d6d-b6fd-4553cc30cae5)

Error in Network Tab
![image](https://github.com/user-attachments/assets/c24d9dfe-665d-4dc5-a1b8-37259a9499bc)

**Comment by fredmaggiowski** [comment]
Hi @rmuller-vu 

Assuming you are using Enhanced Workflow + External Orchestrator Extension API, I've created a project with such setup but managed to successfully generate the configurations with the static replicas value using interpolation.

The error message smells like a bug, so I'd try and go deeper with the hunt; can you please provide further information for helpiung with the repro?

The following info should be enough:

- Console version
- payload your orchestrator is receiving (the part around the value should be enough)
- payload your orchestrator is replying (same as above)
- errors logs of the `backend` and `deploy-service` (if you have access to them, not required but could help a little)

If by any chance you don't feel like this is the proper place for sharing such information please reach out to your Mia-Platform referent.

**Reply by rmuller-vu** [reply]
Sorry for the late reply. This is for Mia Platform v13.0.2 and Console v1.102.8. When we try to deploy with the above scenario our orchestration api doesn't receive any traffic from Mia.

**Reply by rmuller-vu** [reply]
Below is the error message in `backend`. I do not see any errors in `deploy-service`

```json
{"level":50,"time":1725570496682,"pid":6,"hostname":"backend-6dfcbcf896-7lpmf","reqId":"5d728d74-0f32-4d96-9901-82f284211aee","error":{"type":"SyntaxError","message":"Unexpected token u in JSON at position 0","stack":"SyntaxError: Unexpected token u in JSON at position 0\n    at JSON.parse (<anonymous>)\n    at Object.loadConsoleUser (/home/node/app/lib/load-console-user.js:12:37)\n    at Object.loadConsoleUserHook (/home/node/app/backendPlugin.js:98:39)\n    at hookIterator (/home/node/app/node_modules/fastify/lib/hooks.js:293:10)\n    at next (/home/node/app/node_modules/fastify/lib/hooks.js:185:16)\n    at handleResolve (/home/node/app/node_modules/fastify/lib/hooks.js:196:5)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)"},"isUserPropertiesHeaderEmpty":true,"msg":"user is not logged, invalid user properties header"}
```

**Reply by rmuller-vu** [reply]
Any more information you need to reproduce the error with the version we are using?

**Reply by fredmaggiowski** [reply]
Sorry I missed your reply. To be honest, the error log doesn't appear to be related to the issue you are facing. 

Today I'll try to reproduce it again with the version you are using and come back to you as soon as possible.

**Reply by rmuller-vu** [reply]
Were you able to reproduce the issue?

**Reply by fredmaggiowski** [reply]
Sorry @rmuller-vu, I haven't got any time yet. We tracked it internally and will do asap

**Comment by GiovannaMonti** [comment]
Hi @rmuller-vu, we are currently working on resolving the issue.

The error message you're seeing (_Cannot read properties of undefined (reading 'error')_) is likely due to a programming issue on our backend that has now been fixed. Concerning your specific case, your orchestrator generator provider might be failing to generate the files or might not be returning the values to interpolate as our backend expects.

Could you please share the response from the generation API (at least the section containing the public variable value that should be interpolated), or the returned error?
Could you also please share the request payload that is sent to your custom generator (at least the part containing the public variable value)?

**Reply by rmuller-vu** [reply]
Our orchestrator never receives a call from Mia.

**Reply by GiovannaMonti** [reply]
Hi @rmuller-vu,
we have conducted several tests and were able to successfully release projects using the External Orchestrator Extension API with interpolated replicas; we were therefore unable to reproduce your error based on replica interpolation.

However, we did encounter the same error you're seeing in a case where we had made a mistake configuring the provider (we had inserted an incorrect base URL). It might be helpful to check if something similar is happening in your case.

In any case, we’ve also improved the error message to make it more informative.

**Reply by rmuller-vu** [reply]
We are getting successful orchestration if we don't use the interpolation in the replicas field. We only see the error when we try to interpolate the replicas. In this scenario, we don't see any traffic hitting the orchestrator. This is why we believe something happens in Mia that makes it not reach out to the orchestrator.

**Reply by rmuller-vu** [reply]
We are upgrading to a newer version of the console soon. I will re-test when that happens and let you know if we are still having issues

**Reply by fredmaggiowski** [reply]
Thank you @rmuller-vu, if the problem persists please reach out your Mia-Platform referent so that we can schedule a meeting to repro the issue.

**Reply by rmuller-vu** [reply]
We have updated our Mia instance

MIa-Platform: `v13.2.0`
Console Version: `1.108.2`

We are still seeing the error. When I try to use public variable interpolation for a microservice static replicas the deploy fails. I still see no evidence that Mia is reaching out to our orchestrator. This appears to be failing in Mia before it calls the orchestrator.

![image](https://github.com/user-attachments/assets/17804d30-468b-4419-a52d-5b0db96bf409)

**Reply by fredmaggiowski** [reply]
Hi @rmuller-vu thanks for the update, as mentioned I think it would be best if we schedule a meeting to see this live as we couldn't find a way to replicate the issue. Please reach out to your Mia-Platform referent so that we can investigate it on your installation

**Reply by jdloyd** [reply]
Hey quick note - I have tried this today and it seemed to work. @rmuller-vu let me know if you want details.
